libcudacxx/include/cuda/ptx

```diff
@@ -0,0 +1,352 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA_PTX
+#define _CUDA_PTX
+
+#include "std/detail/__config"  // Macros
+#include "std/type_traits"      // std::integral_constant
+#include "../nv/target"         // __CUDA_MINIMUM_ARCH__ and friends
+
+/*
+ * The cuda::ptx namespace intends to provide PTX wrappers for new hardware
+ * features and new PTX instructions so that they can be experimented with
+ * before higher-level C++ APIs are designed and developed.
+ *
+ * The wrappers have the following responsibilities:
+ *
+ * - They must prevent any PTX assembler errors, that is:
+ *   - They are defined only for versions of the CUDA Toolkit in which nvcc/ptxas
+ *     actually recognizes the instruction.
+ *   - Sizes and types of parameters are correct.
+ * - They must convert state spaces correctly.
+ * - They adhere to the libcu++ coding standards of using:
+ *   - Double underscores for all parameters, variables
+ *   - _CUDA_VSTD:: namespace for types
+ *
+ * The wrappers should not do the following:
+ *
+ * - Use any non-native types. For example, an mbarrier instruction wrapper
+ *   takes the barrier address as a uint64_t pointer.
+ *
+ * This header is intended for:
+ *
+ * - internal consumption by higher-level APIs such as cuda::barrier,
+ * - outside developers who want to experiment with the latest features of the
+ *   hardware.
+ *
+ */
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA_PTX
+
+/*
+ * Integral constant types and values for
+ *
+ * - .sem
+ * - .space
+ * - .scope
+ *
+ * Skipping some steps in my reasoning: If we want to keep the PTX bindings
+ * relatively stable, and also be able to adapt to additions of semantics,
+ * space, and scope variants of a PTX instruction, then we must be able to add
+ * new overloads of an instruction with .sem, .space, or .scope as type-level
+ * parameters.
+ *
+ */
+
+enum class dot_sem {
```

Comment:
I assume all below is meant to be public API, as it is not __ugly

Reply:
Yes. I have added a comment to mark them public.